File: backend_fastapi/app/models/student.py
from sqlalchemy import Boolean, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.base import TimestampMixin, UUIDPrimaryKeyMixin


class Student(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "students"

    external_id: Mapped[str | None] = mapped_column(String(64), unique=True, nullable=True)
    roll_number: Mapped[str | None] = mapped_column(String(64), unique=True, nullable=True)
    full_name: Mapped[str] = mapped_column(String(128), nullable=False, index=True)
    class_name: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    section: Mapped[str] = mapped_column(String(32), nullable=False, index=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    embeddings = relationship("FaceEmbedding", back_populates="student", cascade="all, delete-orphan")
    attendance_events = relationship("AttendanceEvent", back_populates="student")

    @property
    def embedding_count(self) -> int:
        return len(self.embeddings)

File: flutter_app/lib/core/models/student.dart
class Student {
  const Student({
    required this.id,
    this.externalId,
    this.rollNumber,
    required this.fullName,
    required this.className,
    required this.section,
    required this.isActive,
    this.embeddingCount = 0,
  });

  final String id;
  final String? externalId;
  final String? rollNumber;
  final String fullName;
  final String className;
  final String section;
  final bool isActive;
  final int embeddingCount;

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      id: json['id'] as String,
      externalId: json['external_id'] as String?,
      rollNumber: json['roll_number'] as String?,
      fullName: json['full_name'] as String,
      className: json['class_name'] as String,
      section: json['section'] as String,
      isActive: json['is_active'] as bool? ?? true,
      embeddingCount: json['embedding_count'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'external_id': externalId,
      'roll_number': rollNumber,
      'full_name': fullName,
      'class_name': className,
      'section': section,
      'is_active': isActive,
      'embedding_count': embeddingCount,
    };
  }
}
